### PR TITLE
Jyl

### PR DESCRIPTION
**Pull request by lijinyang0228**
This change is [Reviewable](https://reviewable.io/reviews/minhyukpark/dataspreadbackend/3)

### CODE REVIEW

File: .bash_profile [hunk 1]
```diff
@@ -0,0 +1 @@
+export PATH=${PATH}:/usr/local/mysql/bin
```
Comment (MinhyukPark) [inline .bash_profile]: Why are we including bash profile as part of the PR?

File: .idea/.gitignore [hunk 2]
```diff
@@ -0,0 +1,2 @@
+# Default ignored files
+/workspace.xml
```
Comment (MinhyukPark) [inline .idea/.gitignore]: the entire .idea directory should be included in the top level .gitignore

File: .bash_profile (hunk 1 above)
Comment (lijinyang0228) [reply .bash_profile]: gitignore

File: .idea/.gitignore (hunk 2 above)
Comment (lijinyang0228) [reply .idea/.gitignore]: ok, I 'll do that.

File: .gitignore [hunk 3]
```diff
@@ -1 +1,3 @@
 dataspreadbackend
+.idea
```
Comment (MinhyukPark) [inline .gitignore]: change this to .idea/ and also run 
```
git rm -r --cached .
```

Comment (lijinyang0228) [reply .gitignore]: fixed

File: main.go [hunk 4]
```diff
@@ -1,20 +1,32 @@
 package main
 
 import (
+    "IndexingStructure/DataStructures/DA"
```
Comment (MinhyukPark) [inline main.go]: This is the part that concerns me the most. How would other people gain access to the Indexing structure repository?

File: main.go [hunk 5]
```diff
@@ -28,7 +40,7 @@ type insertColumnRequest struct {
 }
 
 type insertRowRequest struct {
-    RowIndex int
+    Rowindex int
```
Comment (MinhyukPark) [inline main.go]: Why was this line changed to Rowindex?

File: main.go [hunk 6]
```diff
@@ -91,7 +103,9 @@ func updateColumn(w http.ResponseWriter, r* http.Request) {
     updateLastModified(icr.SqlDatabase, icr.SqlTable)
 }
 
-func updateRow(w http.ResponseWriter, r* http.Request) {
+
+func updateRow(w http.ResponseWriter, r* http.Request, da *DA.DynamicArray) {
```
Comment (MinhyukPark) [inline main.go]: So I don't care about coding style too much but having two different pointer conventions in one line I think is confusing. Can we fix da * DA.DynamicArray to da* DA.DynamicArray?

File: main.go [hunk 7]
```diff
@@ -147,7 +163,7 @@ func getRange(w http.ResponseWriter, r* http.Request) {
     var sqldatabase string
     var sqltable string
 
-    json.Unmarshal([]byte(r.Header.Get("Indices")), &indices)
+    json.Unmarshal([]byte(r.Header.Get("Indices")), &ranks) //rank
```
Comment (MinhyukPark) [inline main.go]: If we're changing the internal structure to rank, why are we still using the header name Indices?

File: main.go [hunk 8]
```diff
@@ -156,28 +172,64 @@ func getRange(w http.ResponseWriter, r* http.Request) {
...
-    ra := sqlinterface.RowAccess{
-        Column: "id",
-        Indices: indices,
+    NumOfRow := len(ranks)
+    endOfRank := ranks[NumOfRow-1]
+    fmt.Printf("ranks:\n") //50
+    fmt.Println(ranks, len(ranks))
+    fmt.Println(da.NumItem()) //11
+
+    if ranks[0]+1 > da.NumItem() {
```
Comment (MinhyukPark) [inline main.go]: So if the header Indices is [500, 89, 30] and the da is 30, 89, 500, wouldn't this code compare 500 + 1 > 3 ? and then it would say no value to display?

File: main.go [hunk 9]
```diff
@@ -156,28 +172,64 @@ func getRange(w http.ResponseWriter, r* http.Request) {
...
+        fmt.Println("ranks[0]+1 > da.numitem(), no value to display\n")
+        w.Header().Set("Content-Type", "application/json; charset=UTF-8")
+        w.WriteHeader(http.StatusOK)
+        if err := json.NewEncoder(w).Encode(nil); err != nil {
+            panic(err)
+        }
+    }  else {
+
+        if NumOfRow > da.NumItem() {
```
Comment (MinhyukPark) [inline main.go]: I think this is assuming that the indices header from the request is always sequential? Is this true?

File: main.go [hunk 10]
```diff
@@ -277,18 +329,77 @@ func updateLastModified(sqldatabase string, sqltable string) {
     ioutil.WriteFile("lastmodified" + "-" + sqldatabase + "-" + sqltable + ".dat", body, 0600)
 }
 
+
+func postLogin(w http.ResponseWriter, r* http.Request, da* DA.DynamicArray) {
+    fmt.Printf("Here is postLogin\n")
+    decoder := json.NewDecoder(r.Body)
+    var irr sqlinfo
```
Comment (MinhyukPark) [inline main.go]: Shouldn't we be making a new struct here instead of using insert row request?

File: main.go [hunk 11]
```diff
@@ -156,28 +172,64 @@ func getRange(w http.ResponseWriter, r* http.Request) {
...
+        // fourtonehundredcDb.OptimizeDB(rowToRankMapArr)
+        // rowToRankMapArr := []map[int]int{rowToRankMap}
+        rows := db.GetRows(ra)
+
+        RowsInTable := make([][]string, NumOfRow)
+        for i := 0; i < NumOfRow; i++ {
+            curRow := rows[i]
+            curRowID, _ := strconv.Atoi(curRow[0])
+            curRank := TableMap[curRowID]
+            RowsInTable[curRank] = curRow
```
Comment (MinhyukPark) [inline main.go]: Does this work properly without frontend changes? Doesn't the frontend currently assume that the rank is the first column instead of getting the rank from the actual ordering of the response?

File: main.go [hunk 12]
```diff
@@ -277,18 +329,77 @@ func updateLastModified(sqldatabase string, sqltable string) {
...
 
+
+func postLogin(w http.ResponseWriter, r* http.Request, da* DA.DynamicArray) {
+    fmt.Printf("Here is postLogin\n")
+    decoder := json.NewDecoder(r.Body)
+    var irr sqlinfo
+    decoder.Decode(&irr)
+    fmt.Println(irr.SqlDatabase)
+
+    indexFile := "index" + "-" + irr.SqlDatabase + "-" + irr.SqlTable + "-" + irr.SqlUser + ".csv"
```
Comment (MinhyukPark) [inline main.go]: Doesn't this imply that we no longer support a single human opening up multiple browser windows with one sql credentials, and if so, should we introduce a mutex lock so that there can't be two file descriptors open at the same time? Another question is what happens if the csv file doesn't exist yet?

File: main.go [hunk 13]
```diff
@@ -277,18 +329,77 @@ func updateLastModified(sqldatabase string, sqltable string) {
...
+    rowIDs := da.RangeQuery(1, da.NumItem())
+    var rowIDstr string
+    rowIDstr = strconv.Itoa(rowIDs[0])
+    for i := 1; i < da.NumItem(); i ++ {
+        rowIDstr += ","
+        rowIDstr += strconv.Itoa(rowIDs[i])
+    }
+    ioutil.WriteFile(indexFile, []byte(rowIDstr), 0600)
```
Comment (MinhyukPark) [inline main.go]: Is this O(n) where n is the number of rows? What happens when we load multiple different tables with millions of rows?

**Comment by lijinyang0228** [comment]
___
*[main.go, line 4 at r3](https://reviewable.io/reviews/minhyukpark/dataspreadbackend/3#-Lw_U7RVblBJsOAGx0cg-r3-4:-Lw_UyXu1-TI9_nHf6He:b-4ydvqy) ([raw file](https://github.com/minhyukpark/dataspreadbackend/blob/7439714f87b1c4c84d005e0e5b62f5b0c7a47738/main.go#L4)):*
<details><summary><i>Previously, MinhyukPark (Minhyuk Park) wrote…</i></summary><blockquote>

This is the part that concerns me the most. How would other people gain access to the Indexing structure repository?
</blockquote></details>

ok I'll change it so we can import from SQL interface

**Comment by lijinyang0228** [comment]
___
*[main.go, line 4 at r3](https://reviewable.io/reviews/minhyukpark/dataspreadbackend/3#-Lw_U7RVblBJsOAGx0cg-r3-4:-LweTyRkE_7k9csCmKU3:brb1d1d) ([raw file](https://github.com/minhyukpark/dataspreadbackend/blob/7439714f87b1c4c84d005e0e5b62f5b0c7a47738/main.go#L4)):*
<details><summary><i>Previously, lijinyang0228 (Jinyang Li) wrote…</i></summary><blockquote>

ok I'll change it so we can import from SQL interface
</blockquote></details>

not it import from github

**Comment by lijinyang0228** [comment]
___
*[main.go, line 166 at r3](https://reviewable.io/reviews/minhyukpark/dataspreadbackend/3#-Lw_U7RVblBJsOAGx0cg-r3-166:-LweWTZg3QKzA3MyFg1b:b-sat53m) ([raw file](https://github.com/minhyukpark/dataspreadbackend/blob/7439714f87b1c4c84d005e0e5b62f5b0c7a47738/main.go#L166)):*
<details><summary><i>Previously, MinhyukPark (Minhyuk Park) wrote…</i></summary><blockquote>

If we're changing the internal structure to rank, why are we still using the header name Indices?
</blockquote></details>

already fixed it, both backend and frontend

**Comment by lijinyang0228** [comment]
___
*[main.go, line 225 at r3](https://reviewable.io/reviews/minhyukpark/dataspreadbackend/3#-Lw_U7RVblBJsOAGx0cg-r3-225:-LweWndr2aG0RUI5Z2Us:b-op11p3) ([raw file](https://github.com/minhyukpark/dataspreadbackend/blob/7439714f87b1c4c84d005e0e5b62f5b0c7a47738/main.go#L225)):*
<details><summary><i>Previously, MinhyukPark (Minhyuk Park) wrote…</i></summary><blockquote>

Does this work properly without frontend changes? Doesn't the frontend currently assume that the rank is the first column instead of getting the rank from the actual ordering of the response?
</blockquote></details>

already changed the frontend accordingly

**Comment by lijinyang0228** [comment]
___
*[main.go, line 336 at r3](https://reviewable.io/reviews/minhyukpark/dataspreadbackend/3#-Lw_U7RVblBJsOAGx0cg-r3-336:-Lwe_5xn12FUl8XNQ0MD:b-6d7kdv) ([raw file](https://github.com/minhyukpark/dataspreadbackend/blob/7439714f87b1c4c84d005e0e5b62f5b0c7a47738/main.go#L336)):*
<details><summary><i>Previously, MinhyukPark (Minhyuk Park) wrote…</i></summary><blockquote>

Shouldn't we be making a new struct here instead of using insert row request?
</blockquote></details>

sorry I don't understand

**Comment by MinhyukPark** [comment]
FYI: @apappu9 @lijinyang0228 not assigning as reviewers but you guys can take a look too and see what the code changes are

**Comment by lijinyang0228** [comment]
___
*[main.go, line 336 at r3](https://reviewable.io/reviews/minhyukpark/dataspreadbackend/3#-Lw_U7RVblBJsOAGx0cg-r3-336:-Lwf7Jvj9hMNxG_M-4a-:bem0354) ([raw file](https://github.com/minhyukpark/dataspreadbackend/blob/7439714f87b1c4c84d005e0e5b62f5b0c7a47738/main.go#L336)):*
<details><summary><i>Previously, MinhyukPark (Minhyuk Park) wrote…</i></summary><blockquote>

you named the variable irr but irr was short for insert row request in the previous parts. We should probably we name it to something else.
</blockquote></details>

rename it

**Comment by lijinyang0228** [comment]
___
*[main.go, line 181 at r3](https://reviewable.io/reviews/minhyukpark/dataspreadbackend/3#-Lw_U7RVblBJsOAGx0cg-r3-181:-Lwf8gTd8wwDmIbozUWh:b9z44rb) ([raw file](https://github.com/minhyukpark/dataspreadbackend/blob/7439714f87b1c4c84d005e0e5b62f5b0c7a47738/main.go#L181)):*
<details><summary><i>Previously, MinhyukPark (Minhyuk Park) wrote…</i></summary><blockquote>

If we assume that the header indices are always sequential and increasing and the`da` always contains all the indices, this would still be erroneous since if our tree has 3 items but the indices are 500, 501, and 502 this check would still fail. Right? @lijinyang0228 
</blockquote></details>

yes, this is a bug and I'll fix it. Besides, why the ranks in the header are not sequential? At least now the frontend makes them sequential. Can we always try to ensure they are sequential in the frontend?

**Comment by MinhyukPark** [comment]
___
*[main.go, line 181 at r3](https://reviewable.io/reviews/minhyukpark/dataspreadbackend/3#-Lw_U7RVblBJsOAGx0cg-r3-181:-Lwf9bAK2-C14uIp5VUP:b-sqzbj8) ([raw file](https://github.com/minhyukpark/dataspreadbackend/blob/7439714f87b1c4c84d005e0e5b62f5b0c7a47738/main.go#L181)):*
<details><summary><i>Previously, lijinyang0228 (Jinyang Li) wrote…</i></summary><blockquote>

yes, this is a bug and I'll fix it. Besides, why the ranks in the header are not sequential? At least now the frontend makes them sequential. Can we always try to ensure they are sequential in the frontend?
</blockquote></details>

I think we can ensure that the rank is sequential in the header as long as we set that as an assumption moving forward. There might be a better way of doing this without assuming that the indices are sequential but there are bigger problems to solve now.

**Comment by lijinyang0228** [comment]
___
*[main.go, line 4 at r3](https://reviewable.io/reviews/minhyukpark/dataspreadbackend/3#-Lw_U7RVblBJsOAGx0cg-r3-4:-LwfB2dr8bSmuDXW82Yb:bb7f2mf) ([raw file](https://github.com/minhyukpark/dataspreadbackend/blob/7439714f87b1c4c84d005e0e5b62f5b0c7a47738/main.go#L4)):*
<details><summary><i>Previously, MinhyukPark (Minhyuk Park) wrote…</i></summary><blockquote>

Can you update the make file to include this package?
</blockquote></details>

done

**Comment by lijinyang0228** [comment]
___
*[main.go, line 181 at r3](https://reviewable.io/reviews/minhyukpark/dataspreadbackend/3#-Lw_U7RVblBJsOAGx0cg-r3-181:-LwfBGSCEnte2iYEZxA7:b1241fy) ([raw file](https://github.com/minhyukpark/dataspreadbackend/blob/7439714f87b1c4c84d005e0e5b62f5b0c7a47738/main.go#L181)):*
<details><summary><i>Previously, MinhyukPark (Minhyuk Park) wrote…</i></summary><blockquote>

I think we can ensure that the rank is sequential in the header as long as we set that as an assumption moving forward. There might be a better way of doing this without assuming that the indices are sequential but there are bigger problems to solve now.
</blockquote></details>

If there are only three items in the database, they should be displayed in the first three rows in the spreadsheet. Getting range 500,501,502 will receive nothing, is that right?

**Comment by lijinyang0228** [comment]
___
*[main.go, line 378 at r3](https://reviewable.io/reviews/minhyukpark/dataspreadbackend/3#-Lw_U7RVblBJsOAGx0cg-r3-378:-LwfD_qt739XUMPNKhp_:b-7w6s0r) ([raw file](https://github.com/minhyukpark/dataspreadbackend/blob/7439714f87b1c4c84d005e0e5b62f5b0c7a47738/main.go#L378)):*
<details><summary><i>Previously, MinhyukPark (Minhyuk Park) wrote…</i></summary><blockquote>

Is this O(n) where n is the number of rows? What happens when we load multiple different tables with millions of rows?
</blockquote></details>

Yes, saving a table is O(n).  One file is for one user and one table. I assume one spreadsheet only  opens one table, so a logout only updates one file.

**Comment by lijinyang0228** [comment]
___
*[main.go, line 340 at r3](https://reviewable.io/reviews/minhyukpark/dataspreadbackend/3#-Lw_U7RVblBJsOAGx0cg-r3-340:-LwfG4_b-h7MB2AUfZOY:bvfmm8f) ([raw file](https://github.com/minhyukpark/dataspreadbackend/blob/7439714f87b1c4c84d005e0e5b62f5b0c7a47738/main.go#L340)):*
<details><summary><i>Previously, MinhyukPark (Minhyuk Park) wrote…</i></summary><blockquote>

Doesn't this imply that we no longer support a single human opening up multiple browser windows with one sql credentials, and if so, should we introduce a mutex lock so that there can't be two file descriptors open at the same time? Another question is what happens if the csv file doesn't exist yet?
</blockquote></details>

If csv file does not exist, a simple way is to send rows in the sequence got from the database, and this is also the initial row order. Or we can set the initial row order according to ascending IDs, but this needs a sort.

**Comment by lijinyang0228** [comment]
___
*[main.go, line 340 at r3](https://reviewable.io/reviews/minhyukpark/dataspreadbackend/3#-Lw_U7RVblBJsOAGx0cg-r3-340:-LwfGmcG6ZRc9L2ApIpX:blo8hv3) ([raw file](https://github.com/minhyukpark/dataspreadbackend/blob/7439714f87b1c4c84d005e0e5b62f5b0c7a47738/main.go#L340)):*
<details><summary><i>Previously, lijinyang0228 (Jinyang Li) wrote…</i></summary><blockquote>

If csv file does not exist, a simple way is to send rows in the sequence got from the database, and this is also the initial row order. Or we can set the initial row order according to ascending IDs, but this needs a sort. 
</blockquote></details>

As for single user or multiple users, we can only allow one user to edit a table and others cannot edit it, so we should add a mutex lock, or we can allow multiple users to share their changes and need some synchronization.

**Comment by lijinyang0228** [comment]
@minhyukpark First, I need to select * from table so I can initialize dynamic array when the file does not exist, so could you please add this function into sql-interface ? Second, I don't see how to delete a row in frontend, and I think we should insert a row to a specific position but now the frontend does not let users choose a position. Maybe it should let user to choose a row, right click and add row above or add row below. Now I'll insert the new row to be the last one. Third, how do you remove a row? I don't find the frontend allow row removing.

**Comment by lijinyang0228** [comment]
___
*[main.go, line 181 at r3](https://reviewable.io/reviews/minhyukpark/dataspreadbackend/3#-Lw_U7RVblBJsOAGx0cg-r3-181:-LwgSW8H8q_V3ByDxTtd:bt8ldl1) ([raw file](https://github.com/minhyukpark/dataspreadbackend/blob/7439714f87b1c4c84d005e0e5b62f5b0c7a47738/main.go#L181)):*
<details><summary><i>Previously, lijinyang0228 (Jinyang Li) wrote…</i></summary><blockquote>

If there are only three items in the database, they should be displayed in the first three rows in the spreadsheet. Getting range 500,501,502 will receive nothing, is that right?
</blockquote></details>

I changed the codes, assuming the ranks got from frontend are sequential. The i^th row is only displayed in the i^th position in frontend.

**Comment by lijinyang0228** [comment]
___
*[main.go, line 190 at r3](https://reviewable.io/reviews/minhyukpark/dataspreadbackend/3#-Lw_U7RVblBJsOAGx0cg-r3-190:-LwgSzT-0BpOy21eN7Vf:b-jy9sbp) ([raw file](https://github.com/minhyukpark/dataspreadbackend/blob/7439714f87b1c4c84d005e0e5b62f5b0c7a47738/main.go#L190)):*
<details><summary><i>Previously, MinhyukPark (Minhyuk Park) wrote…</i></summary><blockquote>

I think this is assuming that the indices header from the request is always sequential? Is this true?
</blockquote></details>

Assume they are always sequential

**Comment by lijinyang0228** [comment]
___
*[main.go, line 390 at r6](https://reviewable.io/reviews/minhyukpark/dataspreadbackend/3#-Lx04Sao4B-Lsv5jv0IW:-LxADtocEddWaH61q1iu:b-oock6g) ([raw file](https://github.com/minhyukpark/dataspreadbackend/blob/b94ef7c04508538f5092391452852e29b337509e/main.go#L390)):*
<details><summary><i>Previously, MinhyukPark (Minhyuk Park) wrote…</i></summary><blockquote>

we should add the initialization here right?
</blockquote></details>

yes, added already, please review

**Comment by lijinyang0228** [comment]
___
*[main.go, line 390 at r6](https://reviewable.io/reviews/minhyukpark/dataspreadbackend/3#-Lx04Sao4B-Lsv5jv0IW:-LxBhuE409iXNx524tbA:b-l7brjc) ([raw file](https://github.com/minhyukpark/dataspreadbackend/blob/b94ef7c04508538f5092391452852e29b337509e/main.go#L390)):*
<details><summary><i>Previously, MinhyukPark (Minhyuk Park) wrote…</i></summary><blockquote>

What does `da.SetArrayAfterIni` do? Otherwise the code looks good to me.
</blockquote></details>

It initializes an empty DA with an array.

**Comment by MinhyukPark** [comment]
___
*[main.go, line 390 at r6](https://reviewable.io/reviews/minhyukpark/dataspreadbackend/3#-Lx04Sao4B-Lsv5jv0IW:-LxDVn0h1bvyJGaM4GhV:b1zgf9l) ([raw file](https://github.com/minhyukpark/dataspreadbackend/blob/b94ef7c04508538f5092391452852e29b337509e/main.go#L390)):*
<details><summary><i>Previously, lijinyang0228 (Jinyang Li) wrote…</i></summary><blockquote>

It initializes an empty DA with an array.
</blockquote></details>

what about the second parameter?

**Comment by lijinyang0228** [comment]
___
*[main.go, line 390 at r6](https://reviewable.io/reviews/minhyukpark/dataspreadbackend/3#-Lx04Sao4B-Lsv5jv0IW:-LxFQ9o-0keSjo9SwQzj:b-pev4l0) ([raw file](https://github.com/minhyukpark/dataspreadbackend/blob/b94ef7c04508538f5092391452852e29b337509e/main.go#L390)):*
<details><summary><i>Previously, MinhyukPark (Minhyuk Park) wrote…</i></summary><blockquote>

what about the second parameter?
</blockquote></details>

node size in the tree.
